api/pkg/transformer/types/series/series.go

```diff
@@ -0,0 +1,153 @@
+package series
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-gota/gota/series"
+
+	"github.com/gojek/merlin/pkg/transformer/types/converter"
+)
+
+type Type string
+
+const (
+	String Type = "string"
+	Int    Type = "int"
+	Float  Type = "float"
+	Bool   Type = "bool"
+)
+
+type Series struct {
+	series *series.Series
+}
+
+type contentType struct {
+	hasFloat  bool
+	hasInt    bool
+	hasBool   bool
+	hasString bool
+}
+
+func NewSeries(s *series.Series) *Series {
+	return &Series{s}
+}
+
+func New(values interface{}, t Type, name string) *Series {
+	s := series.New(values, series.Type(t), name)
+	return &Series{&s}
+}
+
+func NewInferType(values interface{}, seriesName string) (*Series, error) {
+	seriesType := detectType(values)
+	seriesValues, err := castValues(values, seriesType)
+	if err != nil {
+		return nil, err
+	}
+
+	return New(seriesValues, seriesType, seriesName), nil
+}
+
+func (s *Series) Series() *series.Series {
+	return s.series
+}
+
+func (s *Series) Type() Type {
+	return Type(s.series.Type())
+}
+
+func detectType(values interface{}) Type {
+	contentType := &contentType{}
+	v := reflect.ValueOf(values)
+	switch v.Kind() {
+	case reflect.Slice:
+		for i := 0; i < v.Len(); i++ {
+			if v.Index(i).Interface() == nil {
+				continue
+			}
+
+			contentType = hasType(v.Index(i).Interface(), contentType)
```

Comment:
what happen if slice has mix type? how do we determine the type?

Reply:
Yes, it's possible, the test is covered in TestSeries_NewInferType.

Reply:
just curious why the contentType using the last value's type? if the data is {1, 1.2, 3, nil} it will infer that data type is string?